src/cmds/restic/cmd_forget.go

```diff
@@ -157,51 +118,69 @@ func runForget(opts ForgetOptions, gopts GlobalOptions, args []string) error {
 		Tags:    opts.KeepTags,
 	}
 
-	if policy.Empty() {
-		return nil
-	}
-
-	// then, load all remaining snapshots
 	snapshots, err := restic.LoadAllSnapshots(repo)
 	if err != nil {
 		return err
 	}
 
-	// group by hostname and dirs
+	// Group snapshots by hostname and dirs.
 	type key struct {
 		Hostname string
-		Dirs     string
+		Paths    string
+		Tags     string
 	}
 
-	snapshotGroups := make(map[key]restic.Snapshots)
+	snapshotGroups := make(map[key][]*restic.Snapshot)
 
 	for _, sn := range snapshots {
-		if opts.Hostname != "" && sn.Hostname != opts.Hostname {
+		if opts.Host != "" && sn.Hostname != opts.Host {
 			continue
 		}
 
 		if !sn.HasTags(opts.Tags) {
 			continue
 		}
 
-		k := key{Hostname: sn.Hostname, Dirs: strings.Join(sn.Paths, ":")}
+		if !sn.HasPaths(opts.Paths) {
+			continue
+		}
+
+		tags := ""
+		if opts.GroupByTags {
+			tags = strings.Join(sn.Tags, ":")
+		}
+		k := key{Hostname: sn.Hostname, Tags: tags, Paths: strings.Join(sn.Paths, ":")}
```

Comment:
How about I make the separator a `,`? Since that is the character pflags separates slices on, I assume there is no (easy) way to enter it from the cmdline. Nah. take it back. it uses the encoding/csv parser, so it understands about quoting. (`--set 'A,",CH"'`)

More thinking required.